source/reference/data-types.txt

```diff
@@ -38,325 +40,495 @@ string or as a ``Date`` object:
-```````````````````````
 
-To return the date as a string, use the ``Date()`` method, as in the
-following example:
+ObjectId
+--------
 
-.. code-block:: javascript
+``mongosh`` provides the ``ObjectId()`` wrapper class around the
+:ref:`objectid` data type. To generate a new ObjectId, use the following
+operation in ``mongosh``:
 
-   var myDateString = Date();
+.. code-block:: javascript
 
-To print the value of the variable, type the variable name in the
-shell, as in the following:
+   new ObjectId
 
-.. code-block:: javascript
+.. seealso::
 
-   myDateString
+   :method:`ObjectId` 
 
-The result is the value of ``myDateString``:
+.. _shell-type-int:
 
-.. code-block:: javascript
+Int32
+-----
 
-   Wed Dec 19 2012 01:03:25 GMT-0500 (EST)
+If a number can be converted to a 32-bit integer, ``mongosh`` will
+store it as ``Int32``. If not, ``mongosh`` defaults to storing the
+number as a ``Double``. Numerical values that are stored as ``Int32``
+in ``mongosh`` would have been stored by default as ``Double`` in the
+``mongo`` shell.
 
-To verify the type, use the ``typeof`` operator, as in the following:
+The :node-api:`Int32() <Int32.html>` constructor can be used to
+explicitly specify 32-bit integers.
 
 .. code-block:: javascript
 
-   typeof myDateString
+      db.types.insertOne(
+                          {
+                              "_id": 1,
+                              "value": Int32("1"),
+                              "expectedType": "Int32" 
+                          }
+                        )
 
-The operation returns ``string``.
+.. warning::
 
-Return ``Date``
-```````````````
+   Default ``Int32`` and ``Double`` types may be stored inconsistently
+   if you connect to the same collection using both ``mongosh`` and the
+   legacy ``mongo`` shell. 
 
-``mongosh`` wraps objects of ``Date`` type with the
-``ISODate`` helper; however, the objects remain of type ``Date``.
+.. _shell-type-long:
 
-The following example uses both the ``new Date()`` constructor and the
-``ISODate()`` constructor to return ``Date`` objects.
+Long
+----
+
+The :node-api:`Long() <Long.html>` constructor can be used to
+explicitly specify a 64-bit integer.
 
 .. code-block:: javascript
 
-   var myDate = new Date();
-   var myDateInitUsingISODateWrapper = ISODate();
+      db.types.insertOne(
+                          {
+                              "_id": 3,
+                              "value": Long("1"),
+                              "expectedType": "Long" 
+                          }
+                        )
 
-You can use the ``new`` operator with the ``ISODate()`` constructor as
-well.
-
-To print the value of the variable, type the variable name in the
-shell, as in the following:
+.. note::
 
-.. code-block:: javascript
+   In the legacy :binary:`~bin.mongo` shell ``NumberLong()`` accepted
+   either a string or integer value. In ``mongosh``, ``NumberLong()``
+   only accepts string values. ``Long()`` accepts integers and strings.
```

Comment:
```suggestion
   only accepts string values. ``Long()`` accepts strings as well.
```

`Long` doesn’t really accept single arbitrary integer values, it accepts *two* integer values representing the low/high bits of the number. Users should always pass strings, practically speaking.

Reply:
Not sure what you mean, Long() currently accepts a single integer value: 

db.types.insertOne({ "a": 1 })
db.types.insertOne({ "a": Long(1) })
db.types.find({"a": {$type: "long"} })
[ { _id: ObjectId("60d648ed5b37cfdf17351b5b"), a: Long("1") } ]

Maybe it shouldn't do that, but it appears to work on regular integers.

Reply:
`Long(1)` is short for `Long(1, 0)`, yes.

People should still either always pass in strings or two integers, not one.

Reply:
Done.